src/bindings/rust/README.md

```diff
@@ -8,30 +8,151 @@
+# Directly depending on elektra-sys is only needed if you need to use the raw bindings
+elektra-sys = { version = "0.9.0", path = "~/git/libelektra/build/src/bindings/rust/elektra-sys" }
+```
+
+If you run `cargo run` and everything builds correctly and prints `Hello, world!`, you can replace the contents of `main.rs` with the examples shown in the next section.
+
 ## Usage
 
+### Key
+
+An example for using a `StringKey`. For a full example, see the [examples](elektra/src/examples.rs).
+
+```rust
+extern crate elektra;
+use elektra::{ReadableKey, StringKey, WriteableKey};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // To create a simple key with a name and value
+    let mut key = StringKey::new("user/test/language")?;
+    key.set_value("rust");
+
+    println!("Key with name {} has value {}", key.name(), key.value());
+
+    Ok(())
+}
+```
+
+Compared to the C-API, there are two distinct key types, `StringKey` and `BinaryKey`. With these, type mismatches such as calling `keyString` on a `BinaryKey` is not possible. The only difference between them is the type of value you can set and get from them. They are only wrappers over the `Key` from the C-API.
+
+Use a `BinaryKey` for setting arbitrary byte values.
+
+```rust
+extern crate elektra;
+use elektra::{BinaryKey, ReadableKey, WriteableKey};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let binary_content: [u8; 7] = [25, 34, 0, 254, 1, 0, 7];
+    let mut key = BinaryKey::new("user/test/rust")?;
+    key.set_value(&binary_content);
+    let read_content = key.value();
+
+    println!(
+        "Key with name {} holds bytes {:?}",
+        key.name(),
+        read_content
+    );
+
+    Ok(())
+}
+```
+
+The functionality of the keys is split into two traits, `ReadableKey` and `WritableKey`, which define methods that only read information from a key, and modify a key, respectively. For example, the method to retrieve metakeys only returns a key that implements `ReadableKey`, which is named `ReadOnly`. The keys returned cannot be modified in accordance to the design.
+
+### KeySet
+
+A KeySet is a set of StringKeys.
+
+- You can create an empty keyset with `new` or preallocate space for a number of keys with `with_capacity`.
+- It has two implementations of the `Iterator` trait, so you can iterate immutably or mutably.
+
+```rust
+extern crate elektra;
+use elektra::{KeyBuilder, KeySet, ReadableKey, StringKey};
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Create a new KeySet with enough preallocated space for 5 keys
+    let mut keyset = KeySet::with_capacity(5);
+
+    // Append some keys
+    keyset.append_key(
+        KeyBuilder::<StringKey>::new("user/sw/app/#1/host")?
+            .value("localhost")
+            .build(),
+    )?;
+    keyset.append_key(
+        KeyBuilder::<StringKey>::new("user/sw/app/#1/port")?
+            .value("8080")
+            .build(),
+    )?;
+
+    // Iterate the keyset
+    keyset.rewind();
```

Comment:
Is this the usual way how iterators work in Rust? Otherwise, it would be better if you use an external iterator as implemented in C++ (using ksAtCursor and simply incrementing the cursor).

Reply:
I changed the iterators s.t. rewind is no longer necessary. Thanks for the hint!